### DISCUSSION TITLE
[DRAFT] Update version in Mix.install/2 scripts

### DISCUSSION THREAD

**Discussion by sheerlox** [unill-io/semantic-release-hex#21] [Ideas]
<sup>Originally posted by **Eiji7** in #20 [November 21, 2023](https://github.com/talent-ideal/semantic-release-hex/issues/20#issuecomment-1820768597)</sup>
> While `Installation` is auto-generated part of `README` and therefore the most common I think that there may be more things to cover, for example usage within `Elixir`'s script file:
> ```elixir
> # normal call
> Mix.install([:my_lib])
> 
> # without optional parentheses
> Mix.install [:my_lib]
> 
> # using a sigil
> Mix.install ~w[my_lib]a
> ```
> By default `mix` takes the `latest` version, but it's also possible to select specific one …
> ```elixir
> # tuple notation
> Mix.install [{my_lib, "~> 1.0"}]
> 
> # shorter keyword notation
> Mix.install my_lib: "~> 1.0"
> ```
> **Note**: In `Elixir` a `Keyword` notation is allowed at the end of `List`, as a last function/macro call argument or as a nth argument as long as it's wrapped in square brackets.
> 
> For more information please take a look at [Mix.install/2](https://hexdocs.pm/mix/Mix.html#install/2) documentation and [Keywords](https://hexdocs.pm/elixir/syntax-reference.html#keywords) section in `Elixir`'s syntax reference documentation.

**Comment by sheerlox** [comment]
@Eiji7 insightful! If I understand correctly, the use case is when repositories provide standalone scripts that install the package upon running, like the [`examples/vision/mnist.exs`](https://github.com/elixir-nx/axon/blob/cd759774448097c9e2374163a4a508f8aa7d65a7/examples/vision/mnist.exs#L2) script from `axon`?

I've found some other ones doing things slightly differently:
- [`running_gpt_with_lora.exs` from `lorax`](https://github.com/wtedw/lorax/blob/5c31316190e245d709937ebb196b42362e35bf25/examples/running_gpt_with_lora.exs#L17): references the `git` repository, but might benefit from using the version directly if there was a way to update it automatically
- [`setup_local.exs` from `ockam`](https://github.com/hargut/ockam/blob/494ff2e411049998c4b4bd001f9e89bdddcadaa0/examples/elixir/get_started/setup_local.exs#L2): loads the package locally (for development purposes)

Matching the same dependency tuple as in `README` inside `Mix.install` instructions in every `.exs` file in the repository should be pretty easy.

While we discuss this issue, I realize it would be valuable to raise a similar issue in [Renovate](https://github.com/renovatebot/renovate), which I don't think is capable of detecting and updating such version references.

**Reply by Eiji7** [reply]
> repositories provide (…)  scripts that install the package upon running

no and no :wink: 

They usually do not "provide" scripts (like a features), but instead share an **example usage**. They are something between short examples in module/function/macro documentation and apps ready to use in production. Please remember that `mix` app usually is not available on production server and that's why it's not recommended to use it in normal app's runtime code.

`Elixir`'s scripts are used for a simplest projects. Due to `Elixir`'s `10x less LOC` rule which is truly working the apps are often tiny even below `100 LOC`. Instead of split those in few files with documentation it's much cleaner to put all the logic in single file and optionally add a few comments.

Also think a bit about it. The typical `Installation` section code is not correct in most cases. That's just a function definition without the project module and rest of contents. That makes perfect sense for sure, but it's not a code that could be simply copy-pasted.

Often use case are web scrapers. They fetch desired data on page and prints or saves them in specific format (like CSV or JSON).
```elixir
#!/bin/env elixir
# Above line together with `chmod +x script.exs` allows to call `./script.exs`

# Mix.install/2 call goes here

defmodule Example do
  def fetch_page do
    # use some https client hex package here and return response body
  end
  
  def fetch_page_data(body) do
    # using web scraping tools like floki we can easily find a specific node and take it's text content
    # finally returning it in desired format
  end
  
  def format_page_data(data) do
    # Enum.map_join(data, "\n", &Enum.join(&1, ","))
    # or
    # Jason.encode!(data)
  end
  
  def use_formatted_data(data) do
    # IO.puts(data)
    # or
    # File.write!("results.csv", data)
  end
end

Example.fetch_page()
|> Example.fetch_page_data()
|> Example.format_page_data()
|> Example.use_formatted_data()
```

With such a script you can copy-paste everything even to `Elixir`'s shell or `livebook` which is much faster than creating a new empty project.

**Reply by sheerlox** [reply]
Got it, thanks for the in-depth explanation!

I just finished implementing the README update feature (release in [`v1.1.0`](https://github.com/talent-ideal/semantic-release-hex/releases/tag/v1.1.0), for now only available on the `@next` channel), and the regexes I built for that should make updating the `Mix.install/2` versions seem like a breeze :wink: 

On a side note, I have a few plans to further improve the tests on the repo, so I'll prioritize working on that first to increase confidence for potential users.

Thanks again for being so involved in making `semantic-release-hex` great!
